examples/vimco.py

```diff
@@ -0,0 +1,370 @@
+l_dec_h1 = batch_norm(
+    DenseLayer(l_dec_h1, num_units=nhidden, name='P_DENSE1'), name='P_')
+l_dec_x_mu = DenseLayer(
+    l_dec_h1, num_units=num_features,
+    nonlinearity=T.nnet.sigmoid, name='P_X_MU')
+
+z_train, z_mu_train = get_output(
+    [l_z, l_mu], sym_x, deterministic=False)
+x_mu_train = get_output(l_dec_x_mu, z_train, deterministic=False)
+
+z_eval, z_mu_eval = get_output(
+    [l_z, l_mu], sym_x, deterministic=True)
+x_mu_eval = get_output(l_dec_x_mu, z_eval, deterministic=True)
+
+x_mu_sample = get_output(l_dec_x_mu, sym_x, deterministic=True)
+
+params = get_all_params([l_dec_x_mu, l_z], trainable=True)
+q_params = [p for p in params if 'Q_' in p.name]
+p_params = [p for p in params if 'P_' in p.name]
+
+
+def get_vimco_baseline(x, combine=log_sum_exp):
+    """
+    Implements Right term in equation 10 from VIMCO paper
+    x should be a 3d matrix in logspace
+    """
+    K_f = T.cast(x.shape[2], 'float32')
+    return T.log(1.0/(K_f-1.0)) + no_norm_log_baseline_matrix_3d_th(
+        x, combine=log_sum_exp)
+
+
+def no_norm_log_baseline_matrix_3d_th(x, combine=log_sum_exp):
+    """
+    Calculate baseline matrix on last dimension of
+    3d matrix
+    """
+    bs, mc, iw = x.shape
+    x_ = x.reshape((-1, iw))
+    x_ = no_norm_log_baseline_matrix_2d_th(x_, combine=combine)
+    return x_.reshape((bs, mc, iw))
+
+
+def no_norm_log_baseline_matrix_2d_th(x, combine=log_sum_exp):
+    """
+    Calculate unnormalized baselines for VIMCO estimator, sum f, i noteq j
+    """
+    bs, iw = x.shape
+    # create mask that corresponds to sum over i exept when i=j
+    inv_mask = T.eye(iw).reshape((iw, 1, iw)).repeat(bs, axis=1)
+
+    # working in log space to we need to subtract the mask
+    x_masked = x.reshape((1, bs, iw)) - inv_mask*1000000.0
+    m = combine(x_masked, axis=2)
+    return m.transpose()
+
+
+def lower_bound(z, z_mu, x_mu, x, eq_samples, iw_samples, epsilon=1e-6):
+    from theano.gradient import disconnected_grad as dg
+    # reshape the variables so batch_size, eq_samples and iw_samples are
+    # separate dimensions
+    z = z.reshape((-1, eq_samples, iw_samples, latent_size))
+    x_mu = x_mu.reshape((-1, eq_samples, iw_samples, num_features))
+
+    # prepare x, z for broadcasting
+    # size: (batch_size, eq_samples, iw_samples, num_features)
+    x = x.dimshuffle(0, 'x', 'x', 1)
+
+    # size: (batch_size, eq_samples, iw_samples, num_latent)
+    z_mu = z_mu.dimshuffle(0, 'x', 'x', 1)
+
+    log_qz_given_x = log_bernoulli(z, z_mu, eps=epsilon).sum(axis=3)
+    z_prior = T.ones_like(z)*np.float32(0.5)
+    log_pz = log_bernoulli(z, z_prior).sum(axis=3)
+    log_px_given_z = log_bernoulli(x, x_mu, eps=epsilon).sum(axis=3)
+
+    # Calculate the LL using log-sum-exp to avoid underflow
+    log_pxz = log_pz + log_px_given_z
+
+    # L is (bs, mc) See definition of L in appendix eq. 14
+    L = log_sum_exp(log_pxz - log_qz_given_x, axis=2) + \
+        T.log(1.0/T.cast(iw_samples, 'float32'))
```

Comment:
I guess this one can also be simplified to `L = log_mean_exp(log_pxz - log_qz_given_x, axis=2)`